src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameOfT.cs

```diff
@@ -141,7 +141,11 @@ public class Frame<TContext> : Frame
                         }
                     }
 
-                    Reset();
+                    // Don't lose request rejection state
+                    if (!_requestRejected)
+                    {
+                        Reset();
```

Comment:
Only my new tests failed because `Reset()` was dropping the error code. You raised a good point; I'll add a test for the scenario in your other comment.